### PR TITLE
Simplify IManager callback API with zip() function

### PR DESCRIPTION
**Pull request by haydenshively**
This API is both simpler (send 0 if you don't want to update, send new list of positions if you do) and more gas efficient.

The only non-intuitive thing is that if your intent is "no longer consider _any_ positions" you _cannot_ do `return zip([0, 0, 0, 0, 0, 0])`, as this is the same as `return 0`, which tells the code not to write to storage.

Instead, you have to do something like `return zip([0xDEAD, 0xDEAD, 0, 0, 0, 0])`. We will need to call this out in the documentation, but it's actually a good thing. Next time the person goes to add positions, it'll be cheaper to edit storage from {non-zero value} to {another non-zero value} than it is to go from {0} to {non-zero value}

* Note: It doesn't have to be `0xDEAD`, it can be any int24. You can even return non-zero values in the other positions if you want, as long as `firstPartOfEachPair == secondPartOfEachPair`. In other words, `return zip([1, 1, 2, 2, 3, 3])` would clear out your positions just as well as the 0xDEAD thing

### CODE REVIEW

File: core/src/Borrower.sol [hunk 1]
```diff
@@ -107,15 +107,12 @@ contract Borrower is IUniswapV3MintCallback {
         if (allowances[1]) TOKEN1.safeApprove(address(callee), type(uint256).max);
 
         packedSlot.isInCallback = true;
-        int24[] memory positions_ = callee.callback(data);
+        int24[] memory positions_ = positions.write(callee.callback(data));
```
Comment (IanWoodard) [inline core/src/Borrower.sol:110]: I was curious about why this was separated in the earlier PR but this makes sense

Comment (IanWoodard) [reply core/src/Borrower.sol:110]: Is there anything that happens between these two calls that could affect this?

File: core/src/libraries/Positions.sol [hunk 2]
```diff
@@ -14,7 +30,7 @@ uint256 constant Q24 = 0x1000000;
  * @param zipped Encoded Uniswap positions. Equivalent to the layout of `int24[6] storage yourPositions`
  * @return positionsOfNonZeroWidth Flattened array of Uniswap positions that may or may not hold liquidity
  */
-function extract(uint144 zipped) pure returns (int24[] memory positionsOfNonZeroWidth) {
+function extract(uint256 zipped) pure returns (int24[] memory positionsOfNonZeroWidth) {
```
Comment (IanWoodard) [inline core/src/libraries/Positions.sol:33]: Why the change from uint144?

File: core/src/libraries/Positions.sol [hunk 3]
```diff
@@ -59,44 +75,37 @@ function extract(uint144 zipped) pure returns (int24[] memory positionsOfNonZero
 }
 
 library Positions {
-    function write(int24[6] storage stor, int24[] memory mem) internal returns (int24[] memory) {
-        // Validate formatting of Uniswap positions
-        uint256 count = mem.length;
+    function write(int24[6] storage positions, uint256 update) internal returns (int24[] memory) {
```
Comment (IanWoodard) [inline core/src/libraries/Positions.sol:78]: fan of the naming

File: core/src/libraries/Positions.sol (hunk 2 above)
Comment (IanWoodard) [reply core/src/libraries/Positions.sol:33]: I assume this is because it is stored as uint256 by Ethereum on the blockchain.

**Review by IanWoodard — APPROVED** [review]
Overall, LGTM! Good work!

File: core/src/libraries/Positions.sol (hunk 2 above)
Comment (haydenshively) [reply core/src/libraries/Positions.sol:33]: Yeah; it's because under the hood everything is in 256 bit chunks, and Yul (the language used in in-line assembly blocks) treats everything as uint256.

File: core/src/Borrower.sol (hunk 1 above)
Comment (haydenshively) [reply core/src/Borrower.sol:110]: It's just because `write` didn't return anything before, so we had to store the callbacks' return value separately. Code in-between is unrelated.
